src/Features/Core/Portable/Extensions/ExtensionMessageHandlerService.cs

```diff
@@ -0,0 +1,358 @@
+    }
+
+    private void RegisterAssembly(Extension extension, string assemblyFileName, AssemblyHandlers? assemblyHandlers)
+    {
+        lock (_lockObject)
+        {
+            // Make sure a call to UnloadCustomMessageHandlersAsync hasn't happened while we relinquished the lock on _lockObject
+            if (!_extensions.TryGetValue(extension.AssemblyFolderPath, out var currentExtension) || !currentExtension.Equals(extension))
+            {
+                throw new InvalidOperationException($"Handlers in {extension.AssemblyFolderPath} were unregistered while loading handlers.");
+            }
+
+            try
+            {
+                if (assemblyHandlers is not null)
+                {
+                    var duplicateHandler = _workspaceHandlers.Keys.Intersect(assemblyHandlers.WorkspaceMessageHandlers.Keys).Concat(
+                    _documentHandlers.Keys.Intersect(assemblyHandlers.DocumentMessageHandlers.Keys)).FirstOrDefault();
+
+                    if (duplicateHandler is not null)
+                    {
+                        assemblyHandlers = null;
+                        throw new InvalidOperationException($"Handler name {duplicateHandler} is already registered.");
+                    }
+
+                    foreach (var handler in assemblyHandlers.WorkspaceMessageHandlers)
+                    {
+                        _workspaceHandlers.Add(handler.Key, handler.Value);
+                    }
+
+                    foreach (var handler in assemblyHandlers.DocumentMessageHandlers)
+                    {
+                        _documentHandlers.Add(handler.Key, handler.Value);
+                    }
+                }
+            }
+            finally
+            {
+                extension.SetAssemblyHandlers(assemblyFileName, assemblyHandlers);
+            }
+        }
+    }
+
+    private sealed class Extension(ExtensionMessageHandlerService extensionMessageHandlerService, IAnalyzerAssemblyLoaderInternal analyzerAssemblyLoader, string assemblyFolderPath)
+    {
+        /// <summary>
+        /// Gets the object that is used to lock in order to avoid multiple calls from the same extensions to load the same assembly concurrently
+        /// resulting in the constructors of the same handlers being called more than once.
+        /// All other concurrent operations, including modifying <see cref="_assemblies"/> are protected by
+        /// <see cref="ExtensionMessageHandlerService._lockObject"/>.
+        /// </summary>
+        private readonly object _assemblyLoadLockObject = new();
+
+        private readonly Dictionary<string, AssemblyHandlers?> _assemblies = new();
+
+        private readonly ExtensionMessageHandlerService _extensionMessageHandlerService = extensionMessageHandlerService;
+
+        public IAnalyzerAssemblyLoaderInternal AnalyzerAssemblyLoader { get; } = analyzerAssemblyLoader;
+
+        public string AssemblyFolderPath { get; } = assemblyFolderPath;
```

Comment:
don't see why these are props.  this is a private type. c an just be readonly fields.